File: src/main/java/de/kilobyte22/voc/api/network/Network.java
package de.kilobyte22.voc.api.network;

public interface Network {

    boolean connect(Node node);

    boolean disconnect(Node node);

    Node getNode(String address);

    Iterable<Node> getNodes();

    Iterable<Node> getNodes(Node reference);

    void sendToAddress(Node source, String target, String name, Object... data);

    void sendToReachable(Node source, String name, Object... data);

    void sendToVisible(Node source, String name, Object... data);

    void registerSniffer(Sniffer sniffer);

}

File: src/main/scala/de/kilobyte22/voc/api/network/Network.scala
package de.kilobyte22.voc.api.network

import de.kilobyte22.voc.api.Persistable

abstract trait Network extends Persistable {
  def connect(node: Node): Boolean

  def disconnect(node: Node): Boolean

  def getNode(address: String): Node

  def getNodes: Iterable[Node]

  def getNodes(reference: Node): Iterable[Node]

  def sendToAddress(source: Node, target: String, name: String, data: AnyRef*)

  def sendToReachable(source: Node, name: String, data: AnyRef*)

  def sendToVisible(source: Node, name: String, data: AnyRef*)

  def registerSniffer(sniffer: Sniffer)

  def unregisterSniffer(sniffer: Sniffer)
}